File: src/test/java/net/slipp/support/jpa/SchemaGeneratorTest.java
package net.slipp.support.jpa;

import static org.junit.Assert.assertTrue;

import java.io.File;

import net.slipp.domain.tag.TaggedHistory;
import net.slipp.ndomain.tag.NTaggedHistory;

import org.apache.commons.io.FileUtils;
import org.hibernate.cfg.Configuration;
import org.hibernate.tool.hbm2ddl.SchemaExport;
import org.junit.Test;

public class SchemaGeneratorTest {
	private static final String SCHEMA_FROM_JAVA = "src/test/resources/schema/create_schema_java.sql";
	private static final String SCHEMA_FROM_SCALA = "src/test/resources/schema/create_schema_scala.sql";

	@Test
	public void generateSchemaAndCompare() throws Exception {
		generate(TaggedHistory.class, SCHEMA_FROM_JAVA);
		generate(NTaggedHistory.class, SCHEMA_FROM_SCALA);

		assertTrue(FileUtils.contentEquals(new File(SCHEMA_FROM_JAVA), new File(SCHEMA_FROM_SCALA)));
	}

	private void generate(Class<?> entity, String outputfile) {
		Configuration cfg = new Configuration();
		cfg.setProperty("hibernate.hbm2ddl.auto", "create");
		cfg.setProperty("hibernate.dialect", "net.slipp.support.jpa.Mysql5BitBooleanDialect");
		
		cfg.addAnnotatedClass(entity);
		
		SchemaExport export = new SchemaExport(cfg);
        export.setDelimiter(";");
        export.setOutputFile(outputfile);
        export.setFormat(true);
        export.execute(true, false, false, false);
	}
}

File: src/test/scala/net/slipp/support/jpa/SchemaGeneratorTest.scala
package net.slipp.support.jpa

import java.io.File
import java.nio.file.{Paths, Files}

import net.slipp.domain.notification.Notification
import net.slipp.domain.qna.{ScoreLike, Attachment, Answer, Question}
import net.slipp.domain.smalltalk.{SmallTalkComment, SmallTalk}
import net.slipp.domain.tag.{Tag, TaggedHistory}
import net.slipp.domain.user.SocialUser
import org.apache.commons.io.FileUtils
import org.hibernate.cfg.Configuration
import org.hibernate.tool.hbm2ddl.SchemaExport
import org.junit.Assert.assertTrue
import org.junit.Test

class SchemaGeneratorTest {
  private val SCHEMA_DIR = "src/test/resources/schema"
  private val SCHEMA_FILE = s"${SCHEMA_DIR}/create_schema.sql"

  @Test def generateSchemaAndCompare {
    if (!Files.exists(Paths.get(SCHEMA_DIR))) {
      Files.createDirectory(Paths.get(SCHEMA_DIR))
    }

    val entities: Array[Class[_]] = Array(
      classOf[SocialUser],
      classOf[Question],
      classOf[Answer],
      classOf[Tag],
      classOf[Attachment],
      classOf[ScoreLike],
      classOf[TaggedHistory],
      classOf[Notification],
      classOf[SmallTalk],
      classOf[SmallTalkComment]
    )

    generate(entities, SCHEMA_FILE)
  }

  private def generate(entities: Array[Class[_]], outputfile: String) {
    val cfg: Configuration = new Configuration
    cfg.setProperty("hibernate.dialect", "net.slipp.support.jpa.Mysql5BitBooleanDialect")
    entities.foreach(c => cfg.addAnnotatedClass(c))
    val export: SchemaExport = new SchemaExport(cfg)
    export.setDelimiter(";")
    export.setOutputFile(outputfile)
    export.setFormat(true)
    export.execute(true, false, false, false)
  }
}
